File: frontend/dashboard/src/api/commands/commands.ts
import { useQuery } from '@urql/vue'
import { createGlobalState } from '@vueuse/core'

import type { GetAllCommandsQuery } from '@/gql/graphql'

import { commandMenuCacheKey } from '@/api/command-menu.js'
import { useMutation } from '@/composables/use-mutation.js'
import { graphql } from '@/gql/gql.js'

export const invalidationKey = 'CommandsInvalidateKey'

export type Command = GetAllCommandsQuery['commands'][0]

export const useCommandsApi = createGlobalState(() => {
	const useQueryCommands = () =>
		useQuery({
			query: graphql(`
				query GetAllCommands {
					commands {
						id
						name
						description
						aliases
						responses {
							id
							commandId
							text
							twitchCategoriesIds
							twitchCategories {
								id
								name
								boxArtUrl
							}
							onlineOnly
							offlineOnly
						}
						cooldown
						cooldownType
						enabled
						visible
						default
						defaultName
						module
						isReply
						keepResponsesOrder
						deniedUsersIds
						allowedUsersIds
						rolesIds
						onlineOnly
						offlineOnly
						enabledCategories
						requiredWatchTime
						requiredMessages
						requiredUsedChannelPoints
						group {
							id
							name
							color
						}
						groupId
						expiresAt
						expiresType
						roleCooldowns {
							id
							commandId
							roleId
							cooldown
						}
					}
				}
			`),
			context: {
				additionalTypenames: [invalidationKey],
			},
			variables: {},
		})

	const useMutationDeleteCommand = () =>
		useMutation(
			graphql(`
				mutation DeleteCommand($id: UUID!) {
					commandsRemove(id: $id)
				}
			`),
			[invalidationKey, commandMenuCacheKey]
		)

	const useMutationCreateCommand = () =>
		useMutation(
			graphql(`
				mutation CreateCommand($opts: CommandsCreateOpts!) {
					commandsCreate(opts: $opts) {
						id
					}
				}
			`),
			[invalidationKey, commandMenuCacheKey]
		)

	const useMutationUpdateCommand = () =>
		useMutation(
			graphql(`
				mutation UpdateCommand($id: UUID!, $opts: CommandsUpdateOpts!) {
					commandsUpdate(id: $id, opts: $opts)
				}
			`),
			[invalidationKey, commandMenuCacheKey]
		)

	return {
		useQueryCommands,
		useMutationDeleteCommand,
		useMutationCreateCommand,
		useMutationUpdateCommand,
	}
})

File: libs/repositories/commands/commands.go
package commands

import (
	"context"
	"time"

	"github.com/google/uuid"
	"github.com/twirapp/twir/libs/repositories/commands/model"
)

type Repository interface {
	GetManyByChannelID(ctx context.Context, channelID string) ([]model.Command, error)
	GetByID(ctx context.Context, id uuid.UUID) (model.Command, error)
	Create(ctx context.Context, input CreateInput) (model.Command, error)
	Delete(ctx context.Context, id uuid.UUID) error
	Update(ctx context.Context, id uuid.UUID, input UpdateInput) (model.Command, error)
}

type CreateInput struct {
	ChannelID                 string
	Name                      string
	Cooldown                  int
	CooldownType              string
	Enabled                   bool
	Aliases                   []string
	Description               string
	Visible                   bool
	IsReply                   bool
	KeepResponsesOrder        bool
	DeniedUsersIDS            []string
	AllowedUsersIDS           []string
	RolesIDS                  []string
	OnlineOnly                bool
	OfflineOnly               bool
	CooldownRolesIDs          []string
	EnabledCategories         []string
	RequiredWatchTime         int
	RequiredMessages          int
	RequiredUsedChannelPoints int
	GroupID                   *uuid.UUID
	ExpiresAt                 *time.Time
	ExpiresType               *string
}

type UpdateInput struct {
	Name                      *string
	Cooldown                  *int
	CooldownType              *string
	Enabled                   *bool
	Aliases                   []string
	Description               *string
	Visible                   *bool
	IsReply                   *bool
	KeepResponsesOrder        *bool
	DeniedUsersIDS            []string
	AllowedUsersIDS           []string
	RolesIDS                  []string
	OnlineOnly                *bool
	OfflineOnly               *bool
	CooldownRolesIDs          []string
	EnabledCategories         []string
	RequiredWatchTime         *int
	RequiredMessages          *int
	RequiredUsedChannelPoints *int
	GroupID                   *uuid.UUID
	ExpiresAt                 *time.Time
	ExpiresType               *string
}
